File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    let prev = [triangle[0][0]];

    for (let i = 1; i < triangle.length; i++) {
        const row = triangle[i];
        const curr = [];

        curr[0] = prev[0] + row[0];  // 맨 왼쪽

        for (let j = 1; j < row.length - 1; j++) {
            curr[j] = Math.max(prev[j - 1], prev[j]) + row[j];
        }

        curr[row.length - 1] = prev[prev.length - 1] + row[row.length - 1];  // 맨 오른쪽

        prev = curr;  // 다음 단계에 사용할 줄 업데이트
    }

    return Math.max(...prev);
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    row_cnt = len(triangle)
    prev = triangle[0]
    cur_idx = 1
    
    while cur_idx < row_cnt:
        cur = triangle[cur_idx]
        temp = [0] * len(cur)
        
        for i in range(len(cur)):
            num = 0
            if i == 0:
                temp[i] = cur[i] + prev[0]
            elif i == len(cur)-1:
                temp[i] = cur[i] + prev[-1]
            else:    
                temp[i] = cur[i] + max(prev[i-1], prev[i])
        
        prev = temp
        cur_idx += 1
    
    max_val = max(prev)
    return max_val